File: scripts/fork_sync.rb
#!/usr/bin/env ruby
# encoding: utf-8

require 'yaml'

# We just extend the String class to display some colors on the terminal
class String
  # colorization
  # from http://stackoverflow.com/questions/1489183/colorized-ruby-output
  def colorize(color_code); "\e[#{color_code}m#{self}\e[0m"; end
  def red; colorize(31); end
  def green; colorize(32); end
  def yellow; colorize(33); end
  def pink; colorize(35); end
end

# this class contains information regarding all the forked
# repositories
class ForkSync
  attr_reader :forks

  def initialize(config_file)
    begin
      @forks = YAML.load_file(config_file)
    rescue
      abort "Can't open yaml file [ #{e.message} ]"
    end
  end

  def execute
    forks.each do |repo_name, repo_fork|
      puts "--- " + "Synchronizing #{repo_name}".green + " ---"
      dir, upstream_repo = get_values(repo_fork)

      ForkRepo.new(dir, upstream_repo).sync_fork
    end
  end

  def get_values(repo_fork)
    abort "No value given for 'dir'" if [ nil, '' ].include?(repo_fork["dir"])
    abort "No value given for 'upstream'" if [ nil, '' ].include?(repo_fork["upstream"])
    [ repo_fork["dir"], "https://github.com/" + repo_fork["upstream"] ]
  end
end

# this class contains details of a single forked repository
class ForkRepo
  attr_reader :dir, :upstream_repo

  def initialize(dir, upstream_repo)
    @dir = dir
    @upstream_repo = upstream_repo
  end

  def sync_fork
    # change to the directory
    cd_to_dir
    puts "Changing to directory: " + "#{dir}".yellow

    # check for upstream remote
    unless upstream_exists?
      puts "Can't find the upstream remote..."
      puts "Adding #{upstream_repo} as an upstream repository..."
      add_upstream_repo
    end

    puts "Fetching from upstream #{upstream_repo}..."
    fetch_from_upstream

    puts "Merging with the upstream #{upstream_repo}..."
    merge_with_upstream
    puts format_output(current_git_status)
    puts
  end

  private

  def cd_to_dir
    begin
      Dir.chdir(File.expand_path(dir))
    rescue
      abort "Can't find the directory: #{dir}...[ #{e.message} ]"
    end
  end

  def upstream_exists?
    `git ls-remote upstream 2>/dev/null`
    $?.success?
  end

  def add_upstream_repo
    `git remote add upstream #{upstream_repo}`
    unless $?.success?
      abort "error adding the remote upstream: #{upstream_repo}..."
    end
  end

  def fetch_from_upstream
    `git fetch upstream`
    unless $?.success?
      abort "error fetching from upstream: #{upstream_repo}..."
    end
  end

  def merge_with_upstream
    `git checkout master 2>/dev/null` # switch to local master
    `git merge upstream/master` # merge with the upstream master
    unless $?.success?
      abort "  error merging with upstream: #{upstream_repo}..."
    end
  end

  def current_git_status
    `git status`
  end

  def format_output(output)
    s = output.split("\n").at(1)
    if s =~ /[0-9]+/
      "  " + s.pink
    else
      "  " + s.yellow
    end
  end
end

###########################################################
# main program
###########################################################
fork_sync = ForkSync.new('.fork_sync.yaml')
fork_sync.execute

File: scripts/fork_sync.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function
import os
import subprocess
import sys
import json
import re

def get_dev_null():
    """Returns writable /dev/null file descriptor"""
    return open(os.devnull, 'w')

class Colors(object):
    """Class used to colorize the text"""
    Red = '\033[31m'
    Green = '\033[32m'
    Yellow = '\033[33m'
    Pink = '\033[35m'
    End = '\033[0m'

    @staticmethod
    def Colorize(*args):
        """Static method used to colorize the text"""
        return args[1] + args[0] + Colors.End

class ForkSync(object):
    """The main application class which is used to synchronize
    the forks with the upstream.
    """
    def __init__(self, config_file):
        with open(config_file) as cf:
            self.config_data = json.load(cf)

    def run(self):
        for repo in self.config_data:
            Fork(**repo).sync()


class Fork(object):
    """This class handle a single forked repository"""
    def __init__(self, **kwargs):
        self.repo = kwargs

    def sync(self):
        try:
            # prints the starting message for the repository
            print(self._getrepo_start_message())

            # the functions below carry out the fork synchronization per repository
            # the functions are just self-explanatory
            self._change_dir()
            self._pull_from_remote()
            self._add_upstream_if_required()
            self._fetch_from_upstream()
            self._merge_with_upstream()

            # takes the git status and prints it with color formatting,
            # so that the user knows whether merged changes should be pushed or not
            print(self._get_status())

        except Exception as e:
            print('error synchronizing with the fork. {}'.format(e), file=sys.stderr)

    def _getrepo_start_message(self):
        try:
            msg = ' working on ' + Colors.Colorize(self.repo['repo_name'], Colors.Yellow) + ' '
            terminal_width = int(subprocess.check_output(['stty', 'size']).split()[1])
            format_spec = '{:-^'+ str(terminal_width - 10) + '}'
        except:
            raise
        else:
            return format_spec.format(msg)

    def _change_dir(self):
        try:
            os.chdir(os.path.expanduser(self.repo['dir']))
            print('changed to directory: {}'.format(self.repo['dir']))
        except:
            raise

    def _pull_from_remote(self):
        print('pulling from origin/master')
        subprocess.check_call(['git', 'pull'],
                stdout=get_dev_null(), stderr=subprocess.STDOUT, close_fds=True)

    def _add_upstream_if_required(self):
        # checks if the upstream repository exists
        print('checking for upstream')
        try:
            subprocess.check_call(['git', 'ls-remote', 'upstream'],
                    stdout=get_dev_null(), stderr=subprocess.STDOUT, close_fds=True)
        except:
            # since upstream repository is not added, add the upstream repo
            print('can not find the upstream remote...')
            print('adding upstream repostiory...')
            self._add_upstream()
        else:
            print('upstream found')

    def _add_upstream(self):
        subprocess.check_call(['git', 'remote', 'add', 'upstream',
            'https://github.com/' + self.repo['upstream_repo']])

    def _fetch_from_upstream(self):
        print('fetching from upstream')
        subprocess.check_call(['git', 'fetch', 'upstream'])

    def _merge_with_upstream(self):
        subprocess.check_call(['git', 'checkout', 'master'],
                stdout=get_dev_null(), stderr=subprocess.STDOUT, close_fds=True)

        print('merging with upstream')
        subprocess.check_call(['git', 'merge', 'upstream/master'],
                stdout=get_dev_null(), stderr=subprocess.STDOUT, close_fds=True)

    def _get_status(self):
        output = subprocess.check_output(['git', 'status']).split('\n')[1]
        return Colors.Colorize(output, Colors.Red) if re.search('[0-9]+', output) \
               else Colors.Colorize(output, Colors.Green)

def main():
    """Creates a ForkSync object and execute the sycnchronization of the forks"""
    try:
        ForkSync('.fork_sync.json').run()
    except Exception as e:
        print('error: {}'.format(e), file=sys.stderr)

if __name__ == "__main__":
    main()
